### ISSUE TITLE
WooCommerce failed to add the automatic `.product` wrapper to the single product block template (theme's template file) when the single product blocks are inside a block pattern

### ISSUE THREAD

**Issue by daviedR** [woothemes/woocommerce#50055]
### Prerequisites

- [X] I have carried out troubleshooting steps and I believe I have found a bug.
- [X] I have searched for similar bugs in both open and closed issues and cannot find a duplicate.

### Describe the bug

WooCommerce automatically add a `.product` wrapper on single product template if any single product block found in the template.

This logic is handled by the `update_single_product_content` function on the `SingleProductTemplate` file.
And then it will call the `add_compatibility_layer` function on the `SingleProductTemplateCompatibility` file.
In this function, WooCommerce tried to detect any single product block in the template content using the `parse_block` function.

The problem with this logic is when a theme defines its own `single-product.html` template and uses a block pattern to contain the single product blocks, the logic won't detect any single product block.

An example of the `single-product.html` in the theme:

```
<!-- wp:template-part {"slug":"header"} /-->
<!-- wp:pattern {"slug":"mytheme/single-product-content"} /-->
<!-- wp:template-part {"slug":"footer"} /-->
```

Because WooCommerce only uses `parse_blocks` on that template content, the pattern is not resolved. So WooCommerce concludes that there is no single product block inside the template. Thus, the `.product` wrapper will never be added on the front end.

Without the `.product` wrapper, all the styling from the WooCommerce CSS is not applied to the front end, because WooCommerce CSS uses this selector to target single product elements: `.woocommerce div.product ...`.

Note:
If the theme's template is already edited via Site Editor and saved to database, this logic will work, because the saved template content revealed all the single product blocks inside the block pattern, not the block `<!--wp:pattern {...} /-->` call.

### Expected behavior

WooCommerce should resolve the `mytheme/single-product-content` block pattern and detect any single product block inside the pattern content.

### Actual behavior

WooCommerce failed to detect any single product block inside the `mytheme/single-product-content` block pattern.

### Steps to reproduce

1. Create a `single-product.html` template in the theme.
2. Do not put any single product blocks inside the template file. Instead, create a block pattern with single product blocks, and then call the block pattern inside the template file.
3. Check the front end HTML markup, there is no `.product` wrapper generated.

### WordPress Environment

<details><summary>System Status Report</summary>

```
### WordPress Environment ###

WordPress address (URL): [Redacted]
Site address (URL): [Redacted]
WC Version: 9.1.2
Legacy REST API Package Version: The Legacy REST API plugin is not installed on this site.
Action Scheduler Version: ✔ 3.7.4
Log Directory Writable: ✔
WP Version: 6.6
WP Multisite: –
WP Memory Limit: 256 MB
WP Debug Mode: –
WP Cron: ✔
Language: en_US
External object cache: –

### Server Environment ###

Server Info: nginx/1.16.0
PHP Version: 8.3.8
PHP Post Max Size: 1,000 MB
PHP Time Limit: 1200
PHP Max Input Vars: 4000
cURL Version: 8.4.0
(SecureTransport) LibreSSL/3.3.6

SUHOSIN Installed: –
MySQL Version: 8.0.16
Max Upload Size: 300 MB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ✔
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Database ###

[REDACTED]

### Post Type Counts ###

attachment: 24
page: 19
post: 6
product: 18
product_variation: 7
revision: 1
shop_coupon: 1
shop_order_placehold: 1
wp_font_face: 2
wp_font_family: 1
wp_global_styles: 7
wp_navigation: 1
wp_template: 5
wp_template_part: 2
wpcode: 4

### Security ###

Secure connection (HTTPS): ✔
Hide errors from visitors: ✔

### Active Plugins (1) ###

WooCommerce: by Automattic – 9.1.2

### Inactive Plugins (23) ###

Akismet Anti-spam: Spam Protection: by Automattic - Anti-spam Team – 5.3.3
Gutenberg: by Gutenberg Team – 18.8.0
WooPayments: by WooCommerce – 7.9.2
WordPress Beta Tester: by WordPress Upgrade/Install Team – 3.5.6

### Settings ###

API Enabled: –
Force SSL: –
Currency: GBP (£)
Currency Position: left
Thousand Separator: ,
Decimal Separator: .
Number of Decimals: 2
Taxonomies: Product Types: external (external)
grouped (grouped)
simple (simple)
variable (variable)

Taxonomies: Product Visibility: exclude-from-catalog (exclude-from-catalog)
exclude-from-search (exclude-from-search)
featured (featured)
outofstock (outofstock)
rated-1 (rated-1)
rated-2 (rated-2)
rated-3 (rated-3)
rated-4 (rated-4)
rated-5 (rated-5)

Connected to WooCommerce.com: –
Enforce Approved Product Download Directories: ✔
HPOS feature enabled: ✔
Order datastore: Automattic\WooCommerce\Internal\DataStores\Orders\OrdersTableDataStore
HPOS data sync enabled: –

### Logging ###

Enabled: ✔
Handler: Automattic\WooCommerce\Internal\Admin\Logging\LogHandlerFileV2
Retention period: 30 days
Level threshold: –
Log directory size: 81 KB

### WC Pages ###

Shop base: #7 - /shop/
Cart: #8 - /cart/ -  Contains the woocommerce/cart block
Checkout: #9 - /checkout/ -  Contains the woocommerce/checkout block
My account: #10 - /account/
Terms and conditions: #458 - /terms-and-conditions/

### Theme ###

Name: Twenty Twenty-Four
Version: 1.2
Author URL: https://wordpress.org
Child Theme: ❌ – If you are modifying WooCommerce on a parent theme that you did not build personally we recommend using a child theme. See: How to create a child theme
WooCommerce Support: ❌ Not declared

### Templates ###

Overrides: /.../public/wp-content/plugins/woocommerce/templates/block-notices/error.php
/.../public/wp-content/plugins/woocommerce/templates/block-notices/notice.php
/.../public/wp-content/plugins/woocommerce/templates/block-notices/success.php


### Admin ###

Enabled Features: activity-panels
analytics
product-block-editor
coupons
core-profiler
customize-store
customer-effort-score-tracks
import-products-task
experimental-fashion-sample-products
shipping-smart-defaults
shipping-setting-tour
homescreen
marketing
mobile-app-banner
navigation
onboarding
onboarding-tasks
product-custom-fields
remote-inbox-notifications
remote-free-extensions
payment-gateway-suggestions
shipping-label-banner
subscriptions
store-alerts
transient-notices
woo-mobile-welcome
wc-pay-promotion
wc-pay-welcome-page
launch-your-store

Disabled Features: experimental-blocks
minified-js
pattern-toolkit-full-composability
product-pre-publish-modal
printful
settings
async-product-editor-category-field
product-editor-template-system

Daily Cron: ✔ Next scheduled: 2024-07-21 07:59:40 +07:00
Options: ✔
Notes: 79
Onboarding: completed

### Action Scheduler ###

Complete: 559
Oldest: 2024-06-20 01:52:42 +0000
Newest: 2024-07-20 16:22:49 +0000

Failed: 2
Oldest: 2023-12-28 08:04:25 +0000
Newest: 2024-07-09 05:08:31 +0000

Pending: 1
Oldest: 2024-07-21 16:12:19 +0000
Newest: 2024-07-21 16:12:19 +0000


### Status report information ###

Generated at: 2024-07-20 23:22:54 +07:00
```
</details>

### Isolating the problem

- [X] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [X] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/products/storefront/).
- [X] I can reproduce this bug consistently using the steps above.

**Comment by Aljullu** [comment]
Thanks for filling this issue, @daviedR! I opened a PR that will fix it in https://github.com/woocommerce/woocommerce/pull/50595, in case you want to take a look. :slightly_smiling_face:

**Comment by Aljullu** [comment]
@daviedR heads-up that the fix has been merged and if everything goes as expected it will be included in WooCommerce 9.4.0. :tada: Thanks again for your report!

**Comment by daviedR** [comment]
Thanks @Aljullu,  Looking forward to the release!
